legend-sdlc-generation-file/src/main/java/org/finos/legend/sdlc/generation/artifact/ArtifactGenerationFactory.java

```diff
@@ -0,0 +1,129 @@
+    private final PureModelContextData data;
+
+    ArtifactGenerationFactory(PureModel pureModel, PureModelContextData data, List<PackageableElement> elements)
+    {
+        if (pureModel == null || elements == null)
+        {
+            throw new RuntimeException("Pure model and elements required for artifact generation factory");
+        }
+        this.data = data;
+        this.pureModel = pureModel;
+        this.elements = elements;
+        this.extensions = Lists.mutable.withAll(ServiceLoader.load(ArtifactGenerationExtension.class));
+        this.validateNonDuplicateRootPaths();
+    }
+
+    void validateNonDuplicateRootPaths()
+    {
+        Set<String> extensionRootPaths = Sets.mutable.empty();
+        this.extensions.forEach(extension ->
+        {
+            if (!extensionRootPaths.add(extension.getExtensionRootPath()))
+            {
+                throw new RuntimeException("Extension root paths must be unique. Found duplicate root path:" + extension.getExtensionRootPath());
+            }
+        });
+    }
+
+    public static ArtifactGenerationFactory newFactory(PureModel pureModel, PureModelContextData data, List<PackageableElement> elements)
+    {
+        return new ArtifactGenerationFactory(pureModel, data, elements);
+    }
+
+    public MutableMap<ArtifactGenerationExtension, List<ArtifactGenerationResult>> generate()
+    {
+        if (this.extensions.isEmpty() || this.elements.isEmpty())
+        {
+            return Maps.mutable.empty();
+        }
+        MutableMap<ArtifactGenerationExtension, List<ArtifactGenerationResult>> results = Maps.mutable.empty();
+        for (PackageableElement element : this.elements)
+        {
+            org.finos.legend.pure.m3.coreinstance.meta.pure.metamodel.PackageableElement packageableElement = pureModel.getPackageableElement(element.getPath());
+            for (ArtifactGenerationExtension extension : this.extensions)
+            {
+                if (extension.canGenerate(packageableElement))
+                {
+                    List<Artifact> result = this.generateArtifacts(packageableElement, element, extension);
+                    List<GenerationOutput> outputs = ListIterate.collect(result, artifact -> new GenerationOutput(artifact.content, artifact.path, artifact.format));
+                    this.handleResult(results, new ArtifactGenerationResult(packageableElement, outputs, extension));
+                }
+            }
+        }
+        return results;
+    }
+
+    private void handleResult(MutableMap<ArtifactGenerationExtension, List<ArtifactGenerationResult>> results, ArtifactGenerationResult result)
+    {
+        if (results.containsKey(result.getGenerator()))
+        {
+            List<ArtifactGenerationResult> currentResults = results.get(result.getGenerator());
+            currentResults.add(result);
+        }
+        else
+        {
+            results.put(result.getGenerator(), Lists.mutable.with(result));
+        }
+    }
```

Comment:
This method is unnecessary. It can be replaced with this bit of code: `results.getIfAbsentPut(result.getGenerator(), Lists.mutable::empty).add(result)`